File: mctdh_spec_fft/mctdh_spec_fft.m
#!/usr/bin/octave -qf

% Copyright 2015 Jan von Cosel
%
% This file is part of utility-scripts.
%
% utility-scripts is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, either version 3 of the License, or
% (at your option) any later version.
%
% utility-scripts is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
% GNU General Public License for more details.
%
% You should have recieved a copy of the GNU General Public License
% along with utility-scripts. If not, see <http://www.gnu.org/licenses/>.


format long;

fs2au = 41.34137333656; % conversion between femtoseconds and atomic units of time
c0 = 137.0359998;       % the speed of light in atomic units

% read the data.
arg_list = argv();
filename = arg_list{1};
tau = str2num(arg_list{2}) * fs2au;
offset = str2num(arg_list{3});
input_data = load(filename);

% find out the next largest power of two that we need for our spectrum.
N = 1;
while N < length(input_data)
  N = N * 2;
end

% construct the various times and frequencies needed.
dt = (input_data(2,1) - input_data(1,1)) * fs2au;
nyquist = 2.0 * pi / (2.0 * dt);
time = (0.0 : dt : (N - 1) * dt)';
double_time = (-(N-1)*dt : dt : N * dt)';
df = 2.0 * nyquist / N;
double_df = 2.0 * nyquist / (2*N);
freq = (-nyquist : df : nyquist)';
double_freq = (-nyquist : double_df : nyquist)';

% construct the autocorrelation function with cosine weighting and exponential damping.
corr = zeros(N,1);
corr(1:length(input_data)) = input_data(:,2) + j * input_data(:,3);
corr_damp = corr .* exp(-time / tau) .* cos(pi * time / (2.0 * time(N)));

% construct the mirrored autocorrelation function with cosine weighting and exponential damping
double_corr = zeros(2*N,1);
for a=0:N-1
  double_corr(a+1) = conj(corr(N-a));   % complex conjugate the negative part!!!
  double_corr(2*N-1-a) = corr(N-a);
end
double_corr_damp = double_corr .* cos(pi * double_time / (2.0 * time(N))) .* exp(-abs(double_time) / tau);

% this is the actual fourier transformation.
spec = ifft(corr_damp);
double_spec = ifft(double_corr_damp);

% properly arrange the positive and negative parts of the spectrum.
spec_shift = zeros(N+1,1);
double_spec_shift = zeros(2*N+1,1);
spec_shift(1:N/2) = spec((N/2)+1:N);
spec_shift((N/2)+1:N+1) = spec(1:(N/2)+1);
double_spec_shift(1:N) = double_spec(N+1:2*N);
double_spec_shift(N+1:2*N+1) = double_spec(1:N+1);

% write the results to the output files.
specfile = fopen("spec_out.dat", "w");
autofile = fopen("auto_out.dat", "w");
double_specfile = fopen("spec_out_double.dat", "w");
double_autofile = fopen("auto_out_double.dat", "w");
offset_specfile = fopen("spec_out_offset.dat", "w");
fprintf(autofile, "#         Time [au]        Re(Auto)         Im(Auto)\n");
fprintf(specfile, "# Frequency [au]    Frequency [cm-1]        Re(Spec)          Im(Spec)\n");
fprintf(double_autofile, "#         Time [au]        Re(Auto)         Im(Auto)\n");
fprintf(double_specfile, "# Frequency [au]    Frequency [cm-1]        Re(Spec)          Im(Spec)\n");
fprintf(offset_specfile, "# Frequency [au]        Re(Spec)          Im(Spec)\n");
for a=1:length(time)
  fprintf(autofile, "%20.10f %16.10f %16.10f\n", time(a), real(corr_damp(a)), imag(corr_damp(a)));
end
for a=1:length(freq)
  fprintf(specfile, "%15.10f %17.10f %17.10f\n", freq(a), real(spec_shift(a)), imag(spec_shift(a)));
end
for a=1:length(double_time)
  fprintf(double_autofile, "%20.10f %16.10f %16.10f\n", double_time(a), real(double_corr_damp(a)), imag(double_corr_damp(a)));
end
factor = 4.0 * pi^2 / (3.0 * c0);
for a=1:length(double_freq)
  fprintf(double_specfile, "%15.10f %17.10f %17.10f\n", double_freq(a), real(double_spec_shift(a)), imag(double_spec_shift(a)));
  fprintf(offset_specfile, "%15.10f %17.10f %17.10f %17.10f\n", double_freq(a)+offset, real(double_spec_shift(a))*factor*(double_freq(a)+offset), imag(double_spec_shift(a))*factor*(double_freq(a)+offset), abs(double_spec_shift(a))*factor*(double_freq(a)+offset));
end
fclose(specfile);
fclose(autofile);
fclose(double_specfile);
fclose(offset_specfile);
fclose(double_autofile);

File: mctdh_spec_fft/mctdh_spec_fft.f90
! Copyright 2015 Jan von Cosel
!
! This file is part of utility-scripts.
!
! utility-scripts is free software: you can redistribute it and/or modify
! it under the terms of the GNU General Public License as published by
! the Free Software Foundation, either version 3 of the License, or
! (at your option) any later version.
!
! utility-scripts is distributed in the hope that it will be useful,
! but WITHOUT ANY WARRANTY; without even the implied warranty of
! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
! GNU General Public License for more details.
!
! You should have recieved a copy of the GNU General Public License
! along with utility-scripts. If not, see <http://www.gnu.org/licenses/>.

program mctdh_spec_fft
implicit none

integer,parameter :: dp = selected_real_kind(14)    ! double precision kind type
integer,parameter :: inputio = 1                    ! IO unit for the input file
integer,parameter :: specio = 2                     ! IO unit for the autocorrelation output
integer,parameter :: autoio = 3                     ! IO unit for the spectrum output
real(dp),parameter :: fs2au = 41.34137333656_dp     ! conversion between femtoseconds and atomic units of time
real(dp),parameter :: pi = 3.1415926535897932_dp    ! PI
integer :: nlines                                   ! number of lines (= time points) in the input
integer :: N                                        ! number of time points
integer :: stat                                     ! IO status
integer :: i, m, k                                  ! loop index
real(dp) :: tau                                     ! damping time
real(dp) :: auto_real                               ! dummy variable for the real part of auto
real(dp) :: auto_imag                               ! dummy variable for the imaginary part of auto
real(dp) :: auto_abs                                ! dummy variable for the absolute value of auto
real(dp) :: dt                                      ! the time interval
real(dp) :: df                                      ! the frequency interval
real(dp) :: nyquist                                 ! the Nyquist frequency
real(dp) :: tmax                                    ! the maximum time
real(dp),dimension(:),allocatable :: time           ! the time axis
real(dp),dimension(:),allocatable :: double_time    ! the double time axis
real(dp),dimension(:),allocatable :: double_freq    ! the double frequency axis
complex(dp),dimension(:),allocatable :: auto_inp    ! the autocorrelation function read from the file
complex(dp),dimension(:),allocatable :: auto_double ! the mirrored autocorrelation function
complex(dp),dimension(:),allocatable :: spec        ! the spectrum
complex(dp),dimension(:),allocatable :: spec_shift  ! the shifted spectrum
character(len=50) :: inputfile                      ! input file with the autocorrelation function
character(len=50) :: cmd_arg                        ! dummy string for the command line arguments


call get_command_argument(1, inputfile)
if (len_trim(inputfile) == 0) then
    write(*,*) "ERROR opening input file, ", inputfile
    stop
endif
open(unit=inputio,file=inputfile,status='old',action='read')

call get_command_argument(2, cmd_arg)
read(cmd_arg,*,iostat=stat) nlines
if (stat /= 0) then
    write(*,*) "ERROR reading nlines, status:", stat
    stop
endif

call get_command_argument(3, cmd_arg)
read(cmd_arg,*,iostat=stat) tau
if (stat /= 0) then
    write(*,*) "ERROR reading tau, status:", stat
    stop
endif
tau = tau * fs2au

N = 1
do while (N < nlines)
    N = N * 2
enddo

allocate(time(N))
allocate(double_time(2 * N))
allocate(double_freq((2 * N) + 1))
allocate(auto_inp(N))
allocate(auto_double(2 * N))
allocate(spec(2 * N))
allocate(spec_shift((2 * N) + 1))

! skip the first line of the input:
read(inputio,*)
auto_inp = (0.0_dp, 0.0_dp)
do i = 1, nlines
    read(inputio,*) time(i), auto_real, auto_imag, auto_abs
    auto_inp(i) = dcmplx(auto_real, auto_imag)
enddo
close(inputio)

! calculate the required times and frequencies:
dt = (time(2) - time(1)) * fs2au
tmax = dt * real(N - 1, dp)
nyquist = pi / dt
df = nyquist / real(N, dp)
do i = 1, (2 * N) - 1
    double_time(i) = -tmax + real(i - 1, dp) * dt
enddo
do i = 1, (2 * N) + 1
    double_freq(i) = -nyquist + real(i - 1, dp) * df
enddo

! construct the mirrored autocorrelation function with cosine weighting and exponential damping:
auto_double = (0.0_dp, 0.0_dp)
do i = 0, N - 1
    auto_double(i + 1) = dconjg(auto_inp(N - i))    ! complex conjugate the negative part!!!
    auto_double(2 * N - (i + 1)) = auto_inp(N - i)
enddo
auto_double = auto_double * cos(pi * double_time / (2.0_dp * tmax)) * exp(-abs(double_time) / tau)

! perform discrete fourier transformation:
! (Numerical Recipes p. 497, eq 12.1.7)
do m = 0, (2 * N) - 1
    spec(m + 1) = (0.0_dp, 0.0_dp)
    do k = 0, (2 * N) - 1
        spec(m + 1) = spec(m + 1) + auto_double(k + 1) * exp(2.0_dp * pi * (0.0_dp, 1.0_dp) * real(k*m, dp) / real((2*N)-1, dp))
    enddo
enddo

spec_shift(1:N) = spec(N+1:2*N)
spec_shift(N+1:(2*N)+1) = spec(1:N+1)

! write the results to the output files:
open(unit=autoio,file="auto_out.dat",status='replace',action='write')
open(unit=specio,file="spec_out.dat",status='replace',action='write')

write(autoio,'("# Time [au] Re(Auto) Im(Auto)")')
write(specio,'("# Frequency [au] Re(Spec) Im(Spec)")')

do i = 1, 2 * N
    write(autoio,'(f20.10, 2f16.10)') double_time(i), real(auto_double(i)), aimag(auto_double(i))
enddo
do i = 1, (2 * N) + 1
    write(specio,'(f15.10, 2f17.10)') double_freq(i), real(spec_shift(i)), aimag(spec_shift(i))
enddo
close(autoio)
close(specio)

end program mctdh_spec_fft
